### DISCUSSION TITLE
How far is D2 1.0?

### DISCUSSION THREAD

**Discussion by alixander** [terrastruct/d2#905] [General]
E.g. no more breaking changes.

List will continue to be added as we think of it. cc @nhooyr 

1. Happy with rendering of dagre and elk.
    - We need creative shims for things like container edges for dagre.
    - ~~ELK still has capacity to be fine-tuned I believe.~~
2. Add a force-directed layout engine.
    - This is needed for situations where manual layout positions are desired. Dagre and ELK don't support positioning given their hierarchical algorithms. For `top` and `left` to work outside of TALA, a force-directed layout engine is needed.
3. Feature-complete on making layers/scenarios/steps useful.
    - Be able to make ~~gifs~~, an interactive SVG with a navigation UI, maybe an HTML export.
4. Language features: ~~Vars~~, ~~classes~~, array connections, ~~globs~~, ~~imports~~.
5. Sequence diagram robustness.
6. ~~Customizable fonts, e.g. passing TTF~~ or specifying a Google font

**Comment by nhooyr** [comment]
No syntax changes as well. We have a few changes we want to make there that we haven't gotten to. I think the full list is:

- Remove case insensitivity.
- In sequence diagrams spans and edge groups need to be explicitly declared as such instead of our current confusing rules.
- Remove raw pipe strings in favour of just `'` and `"`.

**Comment by nhooyr** [comment]
Also I don't know if we should do it still given it's a minor change but will be a headache for users but we've also discussed changing the comment leader to `//` instead of `#` as `#` conflicts with protected methods in class syntax.

**Reply by alixander** [reply]
yeah i think it's been shown that `//` won't conflict with anything but `#` will. and i doubt any ppl have `//` as the lead for an ID. i'm +1 to that change for 1.0.

**Comment by alixander** [comment]
embed source in diagrams once stable: https://github.com/terrastruct/d2/issues/1000

**Comment by bo-ku-ra** [comment]
- all flags can be written in d2 source.
  - except for -w, -h, -p, and -v  
- eliminated easily confused words
  - 'shape: class' and 'classess{}'
  - comment out charactor '#'
    and
     shape:class
     \ ```#```foo:bar
  - style.animated and --animate-interval

**Comment by bo-ku-ra** [comment]
- multi-language support for output messages (ex.https://github.com/terrastruct/d2/issues/1064#issuecomment-1475475072)

**Comment by bo-ku-ra** [comment]
Do you think it’s possible to reach D2 1.0?
